src/stretch_kinematics_plugin.cpp

```diff
@@ -0,0 +1,313 @@
+
+  const auto arm_ik_solver = arm_jmg_->getGroupKinematics().first.solver_instance_;
+  Eigen::Isometry3d eigen_ik_pose;
+  tf2::fromMsg(ik_pose, eigen_ik_pose);
+  solution = ik_seed_state;
+  // Notation:
+  // B_T_A: The homogenous transformation of A frame w.r.t. B frame
+  // arm_b: Base frame of the kinematics solver for the arm jmg
+  unsigned int attempt = 0;
+  do
+  {
+    ++attempt;
+    // First we solve for the arm only by fixing the mobile base
+    state_->setJointGroupPositions(mobile_base_jmg_, &solution[mobile_base_index_]);
+    state_->updateLinkTransforms();
+    // Convert the desired pose from the arm + mobile base solver base frame to the arm group base frame
+    const auto arm_b_T_ik_pose = tf2::toMsg(state_->getGlobalLinkTransform(arm_ik_solver->getBaseFrame()).inverse() *
+                                            state_->getGlobalLinkTransform(getBaseFrame()) * eigen_ik_pose);
+    const bool ik_valid = arm_ik_solver->searchPositionIK(arm_b_T_ik_pose, arm_jmg_ik_seed_state, timeout * 0.25,
+                                                          arm_jmg_consistency_limits, arm_jmg_solution,
+                                                          solution_callback, error_code, options);
+    if (ik_valid || options.return_approximate_solution)  // found acceptable solution
+    {
+      std::copy(arm_jmg_solution.begin(), arm_jmg_solution.end(), solution.begin());
+
+      if (!solution_callback.empty())
+      {
+        solution_callback(ik_pose, solution, error_code);
+        if (error_code.val != moveit_msgs::msg::MoveItErrorCodes::SUCCESS)
+          continue;
+      }
+      // solution passed consistency check and solution callback
+      error_code.val = moveit_msgs::msg::MoveItErrorCodes::SUCCESS;
+      RCLCPP_DEBUG_STREAM(LOGGER, "Solved after " << (node_->now() - start_time).seconds() << " < " << timeout
+                                                  << "s and " << attempt << " attempts");
+      return true;
+    }
+
+    // If we failed to find a solution for the arm JMG we solve only for the mobile base joint by fixing the arm JMG
+    if (attempt == 1)
+      state_->setJointGroupPositions(arm_jmg_, &arm_jmg_ik_seed_state[0]);
+    else
+      state_->setJointGroupPositions(arm_jmg_, &arm_jmg_solution[0]);
+    state_->updateLinkTransforms();
+    // The transformation for the tip frame w.r.t. the mobile base link given the arm's joint values
+    const auto mobile_base_link_T_current =
+        state_->getGlobalLinkTransform(mobile_base_joint_->getChildLinkModel()).inverse() *
+        state_->getGlobalLinkTransform(getTipFrame());
+    // The error for the mobile base link
+    const auto T_mobile_base_link = eigen_ik_pose * mobile_base_link_T_current.inverse();
+
+    Eigen::AngleAxisd rotation(T_mobile_base_link.rotation());
+    Eigen::Map<Eigen::Vector2d>(&solution.at(mobile_base_index_)) = T_mobile_base_link.translation().block<2, 1>(0, 0);
```

Comment:
it looks like you set the x,y position of the base always to 0,0. is that right?

Reply:
No, I'm setting it to the first two rows of  `T_mobile_base_link.translation()` which are the x,y part of the translation, this's equivalent to `T_mobile_base_link.translation()[0:2, 0:1]` in numpy